Implement MacOS and Windows APIs from `hidapi` 0.12 release

KarlK90 opened the pull request:
This PR implements FFI bindings and safe implementations for the following APIs that where added in `hidapi` release 0.12

**MacOS:**
  * `hid_darwin_get_location_id()`
  * `hid_darwin_set_open_exclusive()`
  * `hid_darwin_get_open_exclusive()`
  * `hid_darwin_is_device_open_exclusive()`
 
**Windows:**
   * `hid_winapi_get_container_id()`

The `hid_darwin_set_open_exclusive()` API is also put to use by introducing the `macos-shared-device` feature flag for MacOS so it is possible to have multiple `HidDevice` handles that access the same physical device.

A convenience addition is the implementation of `Debug` for `HidDevice`, this will only print `HidDevice` but allows wrapping types like `HidResult<HidDevice>` into a `dbg!` macro.

KarlK90 committed:
Implement Debug for HidDevice

..this will only print `HidDevice` but allows wrapping types like
HidResult<HidDevice> into a `dbg!` macro.

KarlK90 committed:
Add MacOS and Windows APIs from HidApi 0.12

This adds FFI bindings and safe implementations for the following APIs
that where added in HidApi release 0.12:

MacOS:

 * hid_darwin_get_location_id()
 * hid_darwin_set_open_exclusive()
 * hid_darwin_get_open_exclusive()
 * hid_darwin_is_device_open_exclusive()

Windows:

 * hid_winapi_get_container_id()

KarlK90 committed:
Add macos-shared-device feature flag

Since `hidapi` 0.12 it is possible to open MacOS devices with shared access, so
that multiple [`HidDevice`] handles can access the same physical device. For
backward compatibility this is an opt-in that can be enabled with the
`macos-shared-device` feature flag.

KarlK90 committed:
Apply clippy lints

ruabmbua commented on src/macos.rs:
```diff
@@ -0,0 +1,65 @@
+        let mut location_id: u32 = 0;
+
+        let res =
+            unsafe { ffi::macos::hid_darwin_get_location_id(self._hid_device, addr_of_mut!(location_id)) };
```
There is no reason to use addr_of_mut!()? Could just obtain a mutable reference to the variable.

KarlK90 replied:
Done! You can see the [changes here.](https://github.com/ruabmbua/hidapi-rs/compare/02a62044aaa72eeea404f30415fc6f7fd3d50e18..4dac1a20d7b4be3df307040a980a4abdbb21a87b)

ruabmbua commented on src/windows.rs:
```diff
@@ -0,0 +1,35 @@
+                addr_of_mut!(container_id),
+            );
+
+            if res == -1 {
```
Does not have to be in unsafe block, prefer smaller isolated unsafe blocks. (even if there have to be two)

KarlK90 replied:
Done! You can see the [changes here.](https://github.com/ruabmbua/hidapi-rs/compare/02a62044aaa72eeea404f30415fc6f7fd3d50e18..4dac1a20d7b4be3df307040a980a4abdbb21a87b)

ruabmbua reviewed (changes requested):
Thanks, LGTM

With some notes though, but easy to fix.

KarlK90 committed:
Add docs.rs specific target_os documentation

This will render better visible `Available on MacOS only` badges for the
specific functions.

KarlK90 commented:
I've added docs.rs specific metadata and documentation hints in the latest commit and also hid the `cfg_libusb_only` macro which should be internal(?). The result will look like this:

![image](https://user-images.githubusercontent.com/13887561/201533942-c6227c9f-c416-4a53-a080-7fdf72704b3a.png)

I find that much nicer, although there are now 2 distinct `Impl HidApi` blocks in the documentation...

KarlK90 commented:
@ruabmbua thanks for the swift review, applied changes and added IMHO better platform specific documentation.

ruabmbua commented:
Thanks for the contribution!
